### ISSUE TITLE
load_state, adding/updating attributes

### ISSUE THREAD

**Issue by geekyGoku** [huggingface/accelerate#1744]
### System Info

```Shell
- `Accelerate` version: 0.20.3
- Platform: Linux-5.4.0-139-generic-x86_64-with-glibc2.31
- Python version: 3.10.11
- Numpy version: 1.24.4
- PyTorch version (GPU?): 2.0.1 (True)
- PyTorch XPU available: False
- System RAM: 944.88 GB
- GPU type: NVIDIA A100 80GB PCIe
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: FSDP
        - mixed_precision: bf16
        - use_cpu: False
        - num_processes: 160
        - machine_rank: 0
        - num_machines: 20
        - main_process_ip: 172.27.14.14
        - main_process_port: 8999
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - fsdp_config: {'fsdp_auto_wrap_policy': 'TRANSFORMER_BASED_WRAP', 'fsdp_backward_prefetch_policy': 'BACKWARD_PRE', 'fsdp_offload_params': True, 'fsdp_sharding_strategy': 2, 'fsdp_state_dict_type': 'FULL_STATE_DICT', 'fsdp_transformer_layer_cls_to_wrap': 'T5Block', 'fsdp_limit_all_gathers': True, 'fsdp_use_orig_params': True}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Hi,

I was trying load from previous checkpoint and set new learning rate and betas in optimizer. 
Expected:

```
 if args.resume_training:
        accelerator.load_state(args.checkpoint)
    
    for param_group in optimizer.param_groups:
        param_group["betas"] = (0.9, 0.95)
        print ("setting learning rate to ", args.lr)
        param_group["lr"] = args.lr

```

However, it always defaults to the original learning rate that was set when save_state was called. Is there something that I am missing. 

Also, I was going through the code and found something interesting in load_state code. 
Inside load state, we call

```
  optimizers = []
        if self.distributed_type == DistributedType.FSDP:
            for i, opt in enumerate(self._optimizers):
                logger.info("Loading FSDP Optimizer")
                self.state.fsdp_plugin.load_optimizer(self, opt, self._models[i], input_dir, i)
                logger.info(f"FSDP Optimizer loaded from input dir {input_dir}")
        elif self.distributed_type not in [DistributedType.DEEPSPEED, DistributedType.MEGATRON_LM]:
            optimizers = self._optimizers`
```
Followed by call to:

```
  load_accelerator_state(
            input_dir,
            models,
            optimizers,
            schedulers,
            self.state.process_index,
            self.scaler,
            map_location,
            **load_model_func_kwargs,
        )
```

Which loads the states again:

 ```
# Optimizer states
    for i, opt in enumerate(optimizers):
        optimizer_name = f"{OPTIMIZER_NAME}.bin" if i == 0 else f"{OPTIMIZER_NAME}_{i}.bin"
        input_optimizer_file = os.path.join(input_dir, optimizer_name)
        optimizer_state = torch.load(input_optimizer_file, map_location=map_location)
        optimizers[i].load_state_dict(optimizer_state)
    logger.info("All optimizer states loaded successfully")
```

Any reason why we are calling load_state_dict twice as we are already calling it inside `self.state.fsdp_plugin.load_optimizer(self, opt, self._models[i], input_dir, i)`

Thanks in advance

### Expected behavior

For the first issue, I want to be able to set new hyperparams for optimizer and scheduler after load state. It currently uses the attributes that we set during save_state and doesn't override to newly set parameters.  Regarding the other question, why are we loading the states twice in FSDP case as we are doing the right thing already in `self.state.fsdp_plugin.load_optimizer(self, opt, self._models[i], input_dir, i)`

**Comment by geekyGoku** [comment]
@muellerzr , tagging you here as you recently committed changes to load_state.

**Comment by muellerzr** [comment]
cc @pacman100 on the FSDP

However @geekyGoku save state is meant to load exactly that, the exact same state. You would need to create a new learning rate scheduler and betas for the optimizer yourself, we do not support behavior such as that.

**Comment by pacman100** [comment]
> Any reason why we are calling load_state_dict twice as we are already calling it inside

Doesn't happen because when using FSDP, the optimizer loaded with FSDP specific code isn't added to the `optimizers` list and as such the  `load_accelerator_state` doesn't load that optimizer.

**Comment by geekyGoku** [comment]
Got it, thanks a lot for the clarification @muellerzr , @pacman100

**Comment by legel** [comment]
I had an issue where my DeepSpeed state was basically corrupted, and was not compatible with my latest model.

So "updating" the DeepSpeed state was necessary...  
Eventually I realized I would need to first just recover my model, then reinitialize the state from scratch.

After a lot of digging, I was able to recover the model weights (phew) by first saving the model using normal methods.

It turns out this could be a way to update / change attributes of the accelerator state as well, by re-initializing it.

```    
# load your model into accelerator
model, ... = accelerator.prepare(model, ...)
# load the state that it had before
accelerator.load_state(input_dir='path_to_your_deepspeed_state'))
# extract the DeepSpeed model wrapper away from your model
model_with_weights = accelerator.unwrap_model(model)
# save your model weights as you normally do -- note, if you have multiple parallel instances, you might want to make model name unique for each (e.g. I used Unix Time to microseconds), to avoid overwriting same file
torch.save(model_with_weights.state_dict(), 'some_path/model_name.pth')

# kill the above program
# then restart the program, except this time, load your model as follows
model.load_state_dict(torch.load('some_path/model_name.pth'), strict=False)
# and then *don't* call accelerator.load_state( ) this time around
# now you're free
```
